admin/class-admin-apple-sections.php

```diff
@@ -0,0 +1,364 @@
+		if ( empty( $taxonomy->label ) ) {
+			wp_die( __( 'You specified an invalid mapping taxonomy.', 'apple-news' ) );
+		}
+
+		// Try to get a list of sections.
+		$section_api = new Section( $this->settings );
+		$sections_raw = $section_api->get_sections();
+		if ( empty( $sections_raw ) || ! is_array( $sections_raw ) ) {
+			wp_die( __( 'Unable to fetch a list of sections.', 'apple-news' ) );
+		}
+
+		// Convert sections returned from the API into a key/value pair of id/name.
+		$sections = array();
+		foreach ( $sections_raw as $section ) {
+			if ( ! empty( $section->id ) && ! empty( $section->name ) ) {
+				$sections[ $section->id ] = $section->name;
+			}
+		}
+
+		// Get mappings from settings.
+		$mappings = array();
+		$settings = get_option( self::TAXONOMY_MAPPING_KEY );
+		if ( ! empty( $settings ) && is_array( $settings ) ) {
+			foreach ( $settings as $section_id => $term_ids ) {
+				foreach ( $term_ids as $term_id ) {
+					$term = get_term( $term_id, $taxonomy->name );
+					if ( ! empty( $term->name ) ) {
+						$mappings[ $section_id ][] = $term->name;
+					}
+				}
+			}
+		}
+
+		// Load the partial with the form.
+		include plugin_dir_path( __FILE__ ) . 'partials/page_sections.php';
+	}
+
+	/**
+	 * Register assets for the options page.
+	 *
+	 * @param string $hook
+	 * @access public
+	 */
+	public function register_assets( $hook ) {
+
+		// Only fire for the hook represented by this class.
+		if ( 'apple-news_page_apple-news-sections' !== $hook ) {
+			return;
+		}
+
+		global $wp_scripts;
+
+		// Enqueue styles for this page.
+		$jquery_ui = $wp_scripts->query('jquery-ui-core');
+		wp_enqueue_style(
+			'apple-news-jquery-ui-autocomplete',
+			'//ajax.googleapis.com/ajax/libs/jqueryui/' . $jquery_ui->ver . '/themes/smoothness/jquery-ui.min.css'
+		);
+		wp_enqueue_style(
+			'apple-news-sections-css',
+			plugin_dir_url( __FILE__ ) . '../assets/css/sections.css'
+		);
+
+		// Enqueue scripts for this page.
+		wp_enqueue_script(
+			'apple-news-sections-js',
+			plugin_dir_url( __FILE__ ) . '../assets/js/sections.js',
+			array( 'jquery', 'jquery-ui-autocomplete' )
+		);
+	}
+
+	/**
+	 * A callback for form save on the section-taxonomy mappings form.
+	 *
+	 * @access private
+	 */
+	private function set_section_taxonomy_mappings() {
+
+		// Ensure we got POST data.
+		if ( empty( $_POST ) || ! is_array( $_POST ) ) {
+			return;
+		}
+
+		// Loop through POST data and extract taxonomy mappings.
+		$mappings = array();
+		$taxonomy = self::get_mapping_taxonomy();
+		foreach ( $_POST as $key => $values ) {
+
+			// Ensure the key is in the right format.
+			if ( 0 !== strpos( $key, 'taxonomy-mapping-' ) ) {
+				continue;
+			}
+
+			// Extract the section ID.
+			$section_id = substr( $key, strlen( 'taxonomy-mapping-' ) );
+
+			// Loop over terms and convert to term IDs for save.
+			foreach ( $values as $value ) {
+
+				// Try to get term.
+				$term = get_term_by( 'name', $value, $taxonomy->name );
```

Comment:
This is expensive. It should at least have a switch for VIP that uses `wpcom_vip_get_term_by`